azure_functions_worker/http_v2.py

```diff
@@ -0,0 +1,292 @@
+                         http_response=http_response,
+                         function=function, fi_context=fi_context, args=args)
+        self.is_async = True
+
+
+class SingletonMeta(type):
+    """
+    Metaclass for implementing the singleton pattern.
+    """
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super().__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class HttpCoordinator(metaclass=SingletonMeta):
+    """
+    HTTP coordinator class for managing HTTP v2 requests and responses.
+    """
+    def __init__(self):
+        self._context_references: Dict[str, BaseContextReference] = {}
+
+    def set_http_request(self, invoc_id, http_request):
+        if invoc_id not in self._context_references:
+            self._context_references[invoc_id] = AsyncContextReference()
+        context_ref = self._context_references.get(invoc_id)
+        context_ref.http_request = http_request
+
+    def set_http_response(self, invoc_id, http_response):
+        if invoc_id not in self._context_references:
+            raise KeyError("No context reference found for invocation %s"
+                           % invoc_id)
+        context_ref = self._context_references.get(invoc_id)
+        context_ref.http_response = http_response
+
+    async def get_http_request_async(self, invoc_id):
+        if invoc_id not in self._context_references:
+            self._context_references[invoc_id] = AsyncContextReference()
+
+        await self._context_references.get(
+            invoc_id).http_request_available_event.wait()
+        return self._pop_http_request(invoc_id)
+
+    async def await_http_response_async(self, invoc_id):
+        if invoc_id not in self._context_references:
+            raise KeyError("No context reference found for invocation %s"
+                           % invoc_id)
+
+        await self._context_references.get(
+            invoc_id).http_response_available_event.wait()
+        return self._pop_http_response(invoc_id)
+
+    def _pop_http_request(self, invoc_id):
+        context_ref = self._context_references.get(invoc_id)
+        request = context_ref.http_request
+        if request is not None:
+            context_ref.http_request = None
+            return request
+
+        raise ValueError("No http request found for invocation %s" % invoc_id)
+
+    def _pop_http_response(self, invoc_id):
+        context_ref = self._context_references.get(invoc_id)
+        response = context_ref.http_response
+        if response is not None:
+            context_ref.http_response = None
+            return response
+
+        raise ValueError("No http response found for invocation %s" % invoc_id)
+
+
+def get_unused_tcp_port():
+    # Create a TCP socket
+    tcp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    # Bind it to a free port provided by the OS
+    tcp_socket.bind(("", 0))
```

Comment:
can we do a check after a certain range? like free port after 1000+

Reply:
host shall address this, confirming with them later